### PR TITLE
checkLabels fails when found missing labels

### PR DESCRIPTION
**Pull request by TheoBrigitte**
Towards: https://github.com/giantswarm/giantswarm/issues/24956

This PR make the checkLabels script fail when missing labels are found.

It also:

- changes the missing label message, to describe what is expected
- add error check
- simplify a range

### CODE REVIEW

File: test/hack/checkLabels/main.go
Comment (QuentinBisson) [inline test/hack/checkLabels/main.go:44]: Can this be moved into a function? GetAllPrometheusRulesLabels or smth? nesting is ewww, especially 4 levels of them :D

**Review by QuentinBisson — APPROVED** [review]
Otherwise looks good

File: test/hack/checkLabels/main.go
Comment (QuantumEnigmaa) [reply test/hack/checkLabels/main.go:44]: Me and my poor coding skills are the ones to blame for this !

**Comment by TheoBrigitte** [comment]
CI fails, which is the expected results

image
https://github.com/giantswarm/prometheus-rules/actions/runs/3884324821/jobs/6626737197

**Review by hervenicol — CHANGES_REQUESTED** [review]
I agree with the idea of this PR, but this actually breaks CI (GH actions).

So I think we should either:
- delay this PR until checkLabels passes
- remove this test from CI (from default make target)

**Review by QuantumEnigmaa — COMMENTED** [review]
You based this on the old version of the script. Could you rebase your branch from the one is this PR : https://github.com/giantswarm/prometheus-rules/pull/595 ?

File: test/hack/checkLabels/main.go
Comment (QuentinBisson) [reply test/hack/checkLabels/main.go:44]: I'm not blaming anyone <3

Comment (QuentinBisson) [reply test/hack/checkLabels/main.go:44]: Goosebumps might have been to much :) But let's pair next time :)

Comment (QuantumEnigmaa) [reply test/hack/checkLabels/main.go:44]: No worries ! Sure I'd love to :)

Comment (TheoBrigitte) [reply test/hack/checkLabels/main.go:44]: I get you Quentin, but for such a small script I would leave this as is for now.

**Comment by TheoBrigitte** [comment]
> So I think we should either:
> 
>     * delay this PR until checkLabels passes
> 
>     * remove this test from CI (from default make target)

I agree to wait until we fixed those errors before merging this.



> You based this on the old version of the script. Could you rebase your branch from the one is this PR : #595 ?

I will wait for your PR to be merged first.

File: test/hack/checkLabels/main.go
Comment (QuentinBisson) [reply test/hack/checkLabels/main.go:44]: Well, I would tend to think that that's what the opsctl creator thought at the time :D but ok 👍🏻

**Comment by TheoBrigitte** [comment]
I think this is fine now, but we have to fix those missing labels now before merging this.

File: helm/prometheus-rules/Chart.yaml
Comment (QuantumEnigmaa) [reply helm/prometheus-rules/Chart.yaml]: Indeed, fixed that.

File: helm/prometheus-rules/values.yaml
Comment (QuantumEnigmaa) [reply helm/prometheus-rules/values.yaml]: Yeah, still forgetting to run `make clean` after local testing :/

**Comment by QuentinBisson** [comment]
Have you considered making make clean a git hook?

**Comment by QuantumEnigmaa** [comment]
> Have you considered making make clean a git hook?

That would be a great idea !

**Comment by QuantumEnigmaa** [comment]
Explanation concerning the changes : 
- Before the changes, the script was using the list of all the source matchers from the alertmanager file to check whether these were well defined in the alerting rules regardless of the fact that its corresponding target matchers.
  - This could lead to some edge cases where source matchers would be defined in the alertmanager file but not in the alerting rules because its corresponding target matchers would not be used to trigger an inhibition. 
- After the changes, the script checks only the existence in the alerting rules of a source matcher whose target matcher corresponds to an existing inhibition label in the alerting rules.

File: test/hack/checkLabels/main.go
Comment (hervenicol) [inline test/hack/checkLabels/main.go:22]: Maybe all a link to our doc (which is not yet super clear either, I have to admit):
https://github.com/giantswarm/prometheus-rules/#inhibitions

Also, if `sourceMatchers and targetMatchers are the labels defined in the alertmanager config file`, maybe naming them `am_sourceMatchers` and `am_targetMatchers` would clarify where they come from?

**Comment by QuantumEnigmaa** [comment]
Thanks to the merge of https://github.com/giantswarm/prometheus-rules/pull/620 the test is now passing. Should we merge this PR ?

File: test/hack/checkLabels/main.go
Comment (QuantumEnigmaa) [reply test/hack/checkLabels/main.go:22]: True, I'll add this and change the labels' name

Comment (QuantumEnigmaa) [reply test/hack/checkLabels/main.go]: Done

**Review by hervenicol — APPROVED** [review]
Fine for me!

**Review by QuantumEnigmaa — APPROVED** [review]

File: test/hack/checkLabels/main.go
Comment (TheoBrigitte) [inline test/hack/checkLabels/main.go:25]: Also this variable name is confusing as we are dealing with a lot of target labels in here. Maybe rename this to something like:
```suggestion
const alertRulesPath = "prometheus-rules/templates/alerting-rules"
```

File: test/hack/checkLabels/main.go
Comment (TheoBrigitte) [inline test/hack/checkLabels/main.go:139]:
```suggestion
			target_dir, err := filepath.Abs(output + "/" + dir.Name() + "/" + alertRulesPath)
```

File: test/hack/checkLabels/main.go
Comment (TheoBrigitte) [inline test/hack/checkLabels/main.go:55]: Why are we checking for the value here ?

File: test/hack/checkLabels/main.go
Comment (TheoBrigitte) [inline test/hack/checkLabels/main.go:60]: What is this break for? This lack a comment at least.

File: test/hack/checkLabels/main.go
Comment (TheoBrigitte) [inline test/hack/checkLabels/main.go:179]: I think there is a flaw here, were we check for a single source matcher. I agree this is what we intend to test here, so keeping this is ok, but we should comment that this is a limitation.
Something like:
```
// Limitation: we check the first source matchers found as we expect only source matcher for each target matcher. But there could be multiple and this part could be improved.
```

File: test/hack/checkLabels/main.go
Comment (hervenicol) [reply test/hack/checkLabels/main.go:23]: I agree, and also better naming of variables.
A step has been made towards this when changing `sourceMatchers` to `am_sourceMatchers`, but I guess we can do more.

File: test/hack/checkLabels/main.go
Comment (hervenicol) [reply test/hack/checkLabels/main.go:179]: In the comment, I'd say "we expect at least one" instead of "only one".

Comment (TheoBrigitte) [reply test/hack/checkLabels/main.go:179]: Yeah good point, I changed that in my suggestion

File: test/hack/checkLabels/main.go
Comment (QuantumEnigmaa) [reply test/hack/checkLabels/main.go:25]: Done

File: test/hack/checkLabels/main.go
Comment (QuantumEnigmaa) [reply test/hack/checkLabels/main.go:139]: Done

File: test/hack/checkLabels/main.go
Comment (QuantumEnigmaa) [reply test/hack/checkLabels/main.go:55]: Because the inhibition labels are booleans so if the selected label isn't one (as `cluster_id` for example) then one do not want to take it into consideration.

File: test/hack/checkLabels/main.go
Comment (QuantumEnigmaa) [reply test/hack/checkLabels/main.go:60]: This is to break out of the current loop if the current label does not match the searched one

File: test/hack/checkLabels/main.go
Comment (QuantumEnigmaa) [reply test/hack/checkLabels/main.go:45]: Originally, I had 2 distinct functions but to avoid code duplication (especially with the 2 for loops) I merged them into a single one and had to use a dirty trick to make it work. 
Would you still prefer to have 2 functions even if it means code duplication ?

File: test/hack/checkLabels/main.go
Comment (QuantumEnigmaa) [reply test/hack/checkLabels/main.go:179]: I added the comment.

File: test/hack/checkLabels/main.go
Comment (QuantumEnigmaa) [reply test/hack/checkLabels/main.go:226]: True. Edited.

File: test/hack/checkLabels/main.go
Comment (QuantumEnigmaa) [reply test/hack/checkLabels/main.go:230]: included above.

File: test/hack/checkLabels/main.go
Comment (QuantumEnigmaa) [reply test/hack/checkLabels/main.go:55]: Added a comment

File: test/hack/checkLabels/main.go
Comment (QuantumEnigmaa) [reply test/hack/checkLabels/main.go:23]: Theo's comment added

File: test/hack/checkLabels/main.go
Comment (QuantumEnigmaa) [reply test/hack/checkLabels/main.go:45]: As discussed, function stayed as it is but I added comments to better explain its purpose.

File: test/hack/checkLabels/main.go
Comment (QuantumEnigmaa) [reply test/hack/checkLabels/main.go:60]: Removed break.

**Comment by QuantumEnigmaa** [comment]
Failing test because of UT but not related to this PR so we should ignore it.

**Review by TheoBrigitte — COMMENTED** [review]
:approved: LGTM

File: test/hack/checkLabels/main.go
Comment (QuantumEnigmaa) [reply test/hack/checkLabels/main.go]: Done.

**Pull request merged**
